src/overlays/actors/ovl_Obj_Grass/z_obj_grass.c

```diff
@@ -194,82 +196,80 @@ void func_809A983C(ObjGrass* this, PlayState* play) {
     s32 j;
     s32 x;
     s32 y;
-    f32 temp_f0;
-    ObjGrassStruct1* ptr;
-    ObjGrassStruct1_1* ptr2;
-    ObjGrassStruct2* ptr3;
+    f32 dist;
+    ObjGrassGroup* grassGroup;
     s16 yaw;
 
-    for (i = 0; i < ARRAY_COUNT(D_809AADC0); i++) {
-        D_809AADC0[i] = NULL;
-        D_809AAE10[i] = 422500.0f;
+    for (i = 0; i < ARRAY_COUNT(sNearestGrassElements); i++) {
+        sNearestGrassElements[i] = NULL;
+        sNearestGrassElementsDist[i] = SQ(650.0f);
     }
 
-    for (i = 0; i < ARRAY_COUNT(D_809AADA0); i++) {
-        D_809AADA0[i] = NULL;
-        D_809AADB0[i] = 422500.0f;
+    for (i = 0; i < ARRAY_COUNT(sNearestGrassGroups); i++) {
+        sNearestGrassGroups[i] = NULL;
+        sNearestGrassGroupsDist[i] = SQ(650.0f);
     }
 
-    for (i = 0; i < this->unk_2944; i++) {
-        ptr = &this->unk_144[i];
+    for (i = 0; i < this->activeGrassGroups; i++) {
+        grassGroup = &this->grassGroups[i];
 
-        for (j = 0; j < ptr->unk_FC; j++) {
-            ptr->unk_0C[j].unk_0F &= (u16)~2;
+        for (j = 0; j < grassGroup->count; j++) {
+            grassGroup->elements[j].elem_flags &= (u16)~2;
         }
     }
 
-    for (i = 0; i < this->unk_2944; i++) {
-        ptr = &this->unk_144[i];
-        temp_f0 = Math3D_Vec3fDistSq(&ptr->unk_00, &player->actor.world.pos);
+    for (i = 0; i < this->activeGrassGroups; i++) {
+        grassGroup = &this->grassGroups[i];
+        dist = Math3D_Vec3fDistSq(&grassGroup->homePos, &player->actor.world.pos);
 
-        for (j = 0; j < ARRAY_COUNT(D_809AADB0); j++) {
-            if (temp_f0 < D_809AADB0[j]) {
+        for (j = 0; j < ARRAY_COUNT(sNearestGrassGroupsDist); j++) {
+            if (dist < sNearestGrassGroupsDist[j]) {
                 break;
             }
         }
 
-        if (j < ARRAY_COUNT(D_809AADB0)) {
-            for (x = 2; x >= j; x--) {
-                D_809AADB0[x + 1] = D_809AADB0[x];
-                D_809AADA0[x + 1] = D_809AADA0[x];
+        if (j < ARRAY_COUNT(sNearestGrassGroupsDist)) {
+            for (x = ARRAY_COUNT(sNearestGrassGroupsDist) - 2; x >= j; x--) {
```

Comment:
Why the `x = ARRAY_COUNT(sNearestGrassGroupsDist) - 2`?

Reply:
So the purpose of this section is to insert (group, dist) into the sorted list `sNearestGrassGroups`/`Dist` at index `j`. If `j` is the last index in the list (length - 1), then we can simply insert it directly into the list. But, if `j` is the second to last index (length - 2) or less, then we need to shift everything over by 1 until we reach index `j`. Or in other words, we copy the second to last element to the last element, decrement, copy third last element to second etc. until the `j`th element has been moved to `j+1`

One thing I will say is that I dislike using `ARRAY_COUNT` here instead of just defining a constant list length. I've only done so because it's the convention, but I feel it would make the code better to tie everything together with a single define.

Reply:
Since there's multiple arrays that need to be synced to this size, I think using a define would be fine?